ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java

```diff
@@ -257,6 +256,10 @@ public void handlePut(RestRequest restRequest, RestResponseChannel restResponseC
       } else if (requestPath.matchesOperation(Operations.UNDELETE) && frontendConfig.enableUndelete) {
         // If the undelete is not enabled, then treat it as unrecognized operation.
         undeleteHandler.handle(restRequest, restResponseChannel, (r, e) -> {
+          if (e instanceof RouterException) {
+            e = new RestServiceException(e.getMessage(),
+                RestServiceErrorCode.getRestServiceErrorCode(((RouterException) e).getErrorCode()), true);
```

Comment:
sorry, I realize that setting this option whenever a failure reason needs to be sent creates a lot of duplicate code. Maybe a better solution would be to have a request-level internal key that can be set to true when that request should be allowed to receive error info (for any exception type).

THe approach in this PR should work, though!

Reply:
This is a good idea, updated the PR to use an internal key.